clippy_lints/src/tests_outside_test_module.rs

```diff
@@ -0,0 +1,101 @@
+use clippy_utils::{diagnostics::span_lint_and_note, is_in_cfg_test, is_in_test_function, is_test_module_or_function};
+use rustc_data_structures::sync::par_for_each_in;
+use rustc_hir::{intravisit::FnKind, Body, FnDecl, HirId, ItemKind, Mod};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::{def_id::LocalDefId, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Triggers when a testing function (marked with the `#[test]` attribute) isn't inside a testing module (marked with `#[cfg(test)]`).
+    ///
+    /// ### Why is this bad?
+    ///
+    /// The idiomatic (and more performant) way of writing tests is inside a testing module (flagged with `#[cfg(test)]`), having test functions outside of this module is confusing and may lead to them being "hidden".
+    ///
+    /// ### Example
+    /// ```rust
+    /// #[test]
+    /// fn my_cool_test() {
+    ///     // [...]
+    /// }
+    ///
+    /// #[cfg(test)]
+    /// mod tests {
+    ///     // [...]
+    /// }
+    ///
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// #[cfg(test)]
+    /// mod tests {
+    ///     #[test]
+    ///     fn my_cool_test() {
+    ///         // [...]
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub TESTS_OUTSIDE_TEST_MODULE,
+    restriction,
+    "The test function `my_cool_test` is outside the testing module `tests`."
+}
+
+pub(crate) struct TestsOutsideTestModule {
+    pub test_mod_exists: bool,
+}
+
+impl TestsOutsideTestModule {
+    pub fn new() -> Self {
+        Self { test_mod_exists: false }
+    }
+}
+
+impl_lint_pass!(TestsOutsideTestModule => [TESTS_OUTSIDE_TEST_MODULE]);
+
+impl LateLintPass<'_> for TestsOutsideTestModule {
+    fn check_mod(&mut self, cx: &LateContext<'_>, _: &Mod<'_>, _: HirId) {
```

Comment:
This should probably be `check_crate` to really only do it once and right at the beginning.

Also: What if the test module is in a completely different file, than the test function? This `check_mod` (or `check_crate`) impl doesn't know the difference. So I'm not sure if checking for it even makes sense. For a restriction lint, I would just not check.

Reply:
`check_mod` is necessary for cases like `test/outer/inner/file.rs`. It works like that on Clippy UI tests (It fails with `check_crate` if it's in a test sub-sub-directory like `tests/ui/tests_outside_test_module/file.rs`).

---

Then maybe changing the lint to pedantic? I think it's pretty universal that if a testing module exists, and some tests are outside of that module, they should be in there.

Reply:
Thinking about this more: Is it relevant if there is a `cfg(test)` module in the first place? I would say no. 

If it is good advice to move a test function to such a module, then it is probably also a good idea to create such a test module.

What do you think about removing this check completely?

Reply:
> If it is good advice to move a test function to such a module, then it is probably also a good idea to create such a test module.

You're right. I'm going to remove the check.